cmd/traffic/cmd/manager/internal/watchable/generic.tmpl.go

```diff
@@ -348,10 +348,14 @@ func (tm *MAPTYPE) coalesce(
 	}
 
 	for {
+		done := ctx.Done()
+		if ctx.Err() != nil {
+			shutdown()
+			done = context.Background().Done()
+		}
 		if snapshot.State == nil {
 			select {
-			case <-ctx.Done():
-				shutdown()
+			case <-done:
```

Comment:
In spite of tests passing, I think there are two issues with this:
1. If `tm.close` is closed, there will still be a CPU consuming loop because that select will always be true until something else is randomly chosen (the <-upstream case).
2. If `ctx.Done()` is closed, then that should have higher precedence than `tm.close`. It should indeed just return.